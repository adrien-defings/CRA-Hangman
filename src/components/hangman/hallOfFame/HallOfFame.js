import PropTypes from 'prop-types'
import React from 'react'

import './hallOfFame.css'

const HallOfFame = ({ entries }) => (
  <div className="hofContainer">
    <h5>Hall Of Fame </h5>
    <table className="hallOfFame table">
      <thead className="">
        <tr>
          <th scope="col">#</th>
          <th scope="col">Word</th>
          {/* <th scope="col">Date</th> */}
          <th scope="col">Moves</th>
          <th scope="col">Player</th>
        </tr>
      </thead>
      <tbody>
        {entries.map(({id, guesses, word, player }, index) => (
          <tr key={id}>
            <td className="index">{index + 1}</td>
            <td className="word">{word}</td>
            <td className="guesses">{guesses}</td>
            <td className="player">{player}</td>
          </tr>
        ))}
      </tbody>
    </table>
  </div>
)

HallOfFame.propTypes = {
  entries: PropTypes.arrayOf(
    PropTypes.shape({
      guesses: PropTypes.number.isRequired,
      word: PropTypes.string.isRequired,
      player: PropTypes.string.isRequired,
      date: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
    })
  ).isRequired,
}

export default HallOfFame

// == Internal helpers ==============================================

export const FAKE_HOF = [
  { guesses: 18, word: 'test', player: 'Jane',  date: '10/10/2017', id: 3, },
  { guesses: 23, word: 'test', player: 'Kevin',  date: '11/10/2017', id: 2, },
  { guesses: 31, word: 'test', player: 'Louisa',  date: '06/10/2017', id: 1, },
  { guesses: 48, word: 'test', player: 'Marc',  date: '14/10/2017', id: 0, },
]

const HOF_KEY = '::Memory::HallofFame'

export function saveHOFEntry(entry, onStored) {
  entry.date = new Date().toLocaleDateString()
  entry.id = Date.now()

  const entries = JSON.parse(localStorage.getItem(HOF_KEY) || '[]')
  const insertionPoint = entries.findIndex(
    ({ guesses }) => guesses >= entry.guesses
  )

  if (insertionPoint === -1) {
    entries.push(entry)
  } else {
    entries.splice(insertionPoint, 0, entry)
  }

  localStorage.setItem(HOF_KEY, JSON.stringify(entries))
  onStored(entries)
}