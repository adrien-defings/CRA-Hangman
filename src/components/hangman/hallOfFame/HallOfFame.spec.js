import React from 'react'

import { shallow } from 'enzyme'
import { expect } from 'chai'

import HallOfFame, {FAKE_HOF} from './HallOfFame'

it('renders without crashing', () => {
  const div = document.createElement('div')
  const wrapper = shallow(<HallOfFame entries={FAKE_HOF}/>, div)
  expect(wrapper).to.matchSnapshot()
})