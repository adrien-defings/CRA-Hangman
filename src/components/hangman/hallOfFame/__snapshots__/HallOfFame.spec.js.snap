// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  className="hofContainer"
>
  <h5>
    Hall Of Fame 
  </h5>
  <table
    className="hallOfFame table"
  >
    <thead
      className=""
    >
      <tr>
        <th
          scope="col"
        >
          #
        </th>
        <th
          scope="col"
        >
          Word
        </th>
        <th
          scope="col"
        >
          Moves
        </th>
        <th
          scope="col"
        >
          Player
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="3"
      >
        <td
          className="index"
        >
          1
        </td>
        <td
          className="word"
        >
          test
        </td>
        <td
          className="guesses"
        >
          18
        </td>
        <td
          className="player"
        >
          Jane
        </td>
      </tr>
      <tr
        key="2"
      >
        <td
          className="index"
        >
          2
        </td>
        <td
          className="word"
        >
          test
        </td>
        <td
          className="guesses"
        >
          23
        </td>
        <td
          className="player"
        >
          Kevin
        </td>
      </tr>
      <tr
        key="1"
      >
        <td
          className="index"
        >
          3
        </td>
        <td
          className="word"
        >
          test
        </td>
        <td
          className="guesses"
        >
          31
        </td>
        <td
          className="player"
        >
          Louisa
        </td>
      </tr>
      <tr
        key="0"
      >
        <td
          className="index"
        >
          4
        </td>
        <td
          className="word"
        >
          test
        </td>
        <td
          className="guesses"
        >
          48
        </td>
        <td
          className="player"
        >
          Marc
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
